src/generation.js

```diff
@@ -135,6 +135,86 @@ class WhisperTimeStampLogitsProcessor extends LogitsProcessor {
     }
 }
 
+const getNgrams = (ngramSize, prevInputIds, numHypos) => {
+  const curLen = prevInputIds.length / numHypos;
+  const generatedNgrams = [];
+  for (let i = 0; i < numHypos; ++i) {
+    const genTokens = prevInputIds.slice(i * curLen, (i + 1) * curLen);
+    const ngrams = [];
+    for (let j = 0; j < curLen + 1 - ngramSize; j++) {
+      const ngram = [];
+      for (let k = 0; k < ngramSize; ++k) {
+        ngram.push(genTokens[j + k]);
+      }
+      ngrams.push(ngram);
+    }
+    const generatedNgram = new Map();
+    for (const ngram of ngrams) {
+      const prevNgram = ngram.slice(0, ngram.length - 1);
+      const prevNgramKey = JSON.stringify(prevNgram);
+      const prevNgramValue = generatedNgram.get(prevNgramKey) ?? [];
+      prevNgramValue.push(ngram[ngram.length - 1]);
+      generatedNgram.set(prevNgramKey, prevNgramValue);
+    }
+    generatedNgrams.push(generatedNgram);
+  }
+  return generatedNgrams;
+};
+
+const getGeneratedNgrams = (bannedNgrams, prevInputIds, ngramSize, curLen) => {
+  const ngramIdx = prevInputIds.slice(curLen + 1 - ngramSize, curLen);
+  const banned = bannedNgrams.get(JSON.stringify(ngramIdx)) ?? [];
+  return banned;
+};
+
+const calcBannedNgramTokens = (ngramSize, prevInputIds, numHypos, curLen) => {
+  const batchBannedTokens = [];
+  if (curLen + 1 < ngramSize) {
+    for (let i = 0; i < numHypos; ++i) {
+      batchBannedTokens.push([]);
+    }
+  } else {
+    const generatedNgrams = getNgrams(ngramSize, prevInputIds, numHypos);
+
+    for (let i = 0; i < numHypos; ++i) {
+      const hypoPrevInputIds = prevInputIds.slice(i * curLen, (i + 1) * curLen);
```

Comment:
> curLen is just `input_ids.length` (see comment on main thread for why)

Given what you said ^, this line will have to change to just `prevInputIds` without slicing into it. I was referencing this line in the HuggingFace implementation: https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L468

Reply:
also on this thread, I'm trying to implement the EncoderNoRepeatNGramLogitsProcessor with this change in mind.

Looking at this line in the HuggingFace implementation: https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L530
what would the `self.batch_size` be if we're only getting a flat array?

Reply:
ok, I have an attempt at NoRepeatNGramLogitsProcessor and RepetitionPenaltyLogitsProcessor